pcicrawler/ocp_diag.py

```diff
@@ -0,0 +1,259 @@
+    def run(self):
+        return super().run(self._device.express_link.capable_width)
+
+
+class CheckAddress(OCPTestStep):
+    _test_name = "pci_address_check"
+
+    def run(self):
+        dut = tv.Dut(id=str(self._device), name="PCIe Device")
+        with self._ocp_run.scope(dut=dut):
+            step = self._ocp_run.add_step(self._test_name)
+            if str(self._device) not in str(self._expected_result):
+                step.add_diagnosis(tv.DiagnosisType.FAIL, 
+                                verdict="Test {} did not find {} in expected list {}.".format(
+                                    self._test_name,
+                                    self._device,
+                                    self._expected_result))
+            else:
+                step.add_diagnosis(tv.DiagnosisType.PASS, 
+                                verdict="Test {} for device {} PASSED".format(
+                                    self._test_name,
+                                    self._device))
+
+
+class CheckAER(OCPTestStep):
+    _test_name = "pci_AER_check"
+
+    def __init__(self,
+                    device: PCIDevice,
+                    expected_results: str,
+                    ocp_run: tv.TestRun):\
+        super().__init__(device, expected_results, ocp_run)
+
+    def run(self):
+        return super().run(self._device.express_aer)
+
+
+class TestSet(object):
+    '''Object to hold a list of DUTs to be tested and 
+    the values they need to be compared against.'''
+    def __init__(self, 
+                 duts: ty.List[PCIDevice], 
+                 validate:ty.Dict[str, ty.Union[int, str, ty.Dict[str, str]]]):
+        self.duts = duts
+        self.conditions = validate
+
+    def __str__(self):
+        return str(self.duts)
+
+
+# List of available tests to iterate through
+AVAILABLE_TESTS = {
+    'location': CheckLocation,
+    'vendor_id': CheckVendorID,
+    'device_id': CheckDeviceID,
+    'class_id': CheckClassID,
+    'physical_slot': CheckPhysicalSlot,
+    'current_link_speed': CheckCurrentLinkSpeed,
+    'current_link_width': CheckCurrentLinkWidth,
+    'capable_link_speed': CheckCapableLinkSpeed,
+    'capable_link_width': CheckCapableLinkWidth,
+    'addresses': CheckAddress,
+    'check_aer': CheckAER,
+}
+
+
+class OCPOutputObj:
+    def __init__(self,
+                 devs: ty.Set[PCIDevice],
+                 json_path: str,
+                 ocp_run: tv.TestRun):
+        self._json_path = json_path
+        self._devs = devs
+        self._ocp_run = ocp_run
+        self._result = True
+
+    def run(self):
+        duts = []
+        # Process input json
+        with open(self._json_path, 'r') as fh:
+            input_json = json.load(fh)
+            try:
+                duts = input_json['duts']
+            except KeyError as _:
+                raise json.JSONDecodeError("ERROR: Cannot parse file {}, missing DUTs info.".format(self._json_path))
+        filtered_duts = []
+        for dut in duts:
```

Comment:
why is this filtering logic done here and is not input for this object called `OCPOutputObj`. why is the _output_ deciding what devices to run on?

Reply:
at this point i cant remember what i wanted to say here; i think it was about discoverability of the hardware, ie your selector should overlay on the existing hardware, rather than making a list of pcie devices out of thin air.
i'll close this